torch/distributed/fsdp/_shard_utils.py

```diff
@@ -190,3 +190,24 @@ def _create_chunk_dtensor(
         device_mesh=device_mesh,
         placements=shard_placements,
     )
+
+
+def _all_gather_dtensor(
```

Comment:
Can we merge this logic and the one in `torch/distributed/tensor/parallel/fsdp.py` into one common place?

Reply:
We cannot. This `_all_gather_dtensor` is internal to FSDP.  We are following the extension design here. See this function as an example: https://github.com/pytorch/pytorch/blob/main/torch/distributed/fsdp/_fsdp_extensions.py#L91 

Essentially, there are two code paths:
FSDP only --> _all_gather_dtensor()
FSDP + TP -> _extensions.all_gather_dtensor